### PR TITLE
Bisal Code Files

### PR DESCRIPTION
**Pull request by codewithbisal**
Code files for experiment 1 2 and 3

### CODE REVIEW

File: Bisal/assemblycodes/factorial.s [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    j    end                        
+
+compute_factorial:
+    mul  t0, t0, t1    # t0 = t0 * t1 (multiply current factorial by current counter)
+    mv   t1, t2        # Update t1 to t2 (increment counter)
+    j    check     # Repeat factorial
+
+end:
+    li   a7, 10        # syscall: exit
+    ecall
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/factorial.s:33]: Why is this ecall here and not the exit logic of Spike?

File: Bisal/assemblycodes/factorial.s [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/factorial.s:35]: Remove these extra spaces

File: Bisal/assemblycodes/reverse.s [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    addi t0, t0, 4            # for starting pointer add 4 as each element is of 4 bytes
+    addi t1, t1, -4           # for ending pointer substract 4 as each element is of 4 bytes
+
+    # check loop condition
+    addi t2, t2, -1           
+    bnez t2, reverse_loop     # If t2 != 0, repeat the loop
+
+end_reverse:
+    li a7, 10                 # syscall: exit
+    ecall
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/reverse.s:45]: Same ecall is appearing here. It seems like you have copied the code from somewhere and while pasting, you messed up with the original template which are provided to you.

File: Bisal/assemblycodes/sort.s [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#  Description: File consists of code based for insertion sort algorithm
+#  Author:      Bisal Saeed
+#  Date:        7/12/2024
+#  ============================================================================
+
+.section .text
+.globl _start
+
+_start:
+    .data
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/sort.s:12]: What is the purpose of this directive? It should not be here

File: Bisal/assemblycodes/sort.s [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    .align 2
+array:
+    .word 5, 2, 4, 6, 1, 3    # Array to be sorted
+
+array_size = (. - array) / 4  # Calculate array size in words
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/sort.s:17]: This is not a standard assembly syntax and all compilers does not support it.

File: Bisal/assemblycodes/sort.s [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    bge  t1, array_size, end_sort  
+
+    # Inner loop
+    mv   t2, t1           # t2 = i (index of element to be inserted)
+    lw   t3, 0(t0)        # Load array[i] into t3
+    li   t4, 0            # Initialize (index for comparison)
+
+inner_loop:
+    # Exit loop if j >= i 
+    blt  t4, t2, check_next
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/sort.s:34]: Where is this "check_next" label? It is not in this file so this line will generate an error. You are supposed to commit a working and error free code.

File: Bisal/assemblycodes/sort.s [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+shift_end:
+    sw   t3, 0(t0 + t4*4)  # Store array[i] into array[j] (correct position )
+    addi t1, t1, 1         
+    j    outer_loop        
+
+end_sort:
+    # Exit program
+    li   a7, 10            # syscall: exit
+    ecall
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/sort.s:61]: Once again here is an ecall. I am getting an impression that this code was generated via GPT.

File: Bisal/assemblycodes/sort.s (hunk 6 above)
Comment (UmerShahidengr) [inline Bisal/assemblycodes/sort.s:34]: This logic is wrong, logic is not matching with the comment, it is checking j<i but comment is saying that you are checking if j>=i.

File: Bisal/assemblycodes/task1_2.s [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    # for bitcount
+    li a3,12
+    # temporary registers
+    li t1, 0            # Initialize count of set bitss to 0 
+    li t0, 32           # Bit counter upto 32 bits
+    
+
+# calculate difference between two numbers 
+diff:
+    # a0-a1=a2
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/task1_2.s:22]: inverted comment, comment should be a2=a0-a1

File: Bisal/assemblycodes/task1_2.s [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    sub a2,a0,a1 
+    # check if x2 is smaller than x3 for absolute difference -->branch to done
+    bge a0,a1,done
```
Comment (UmerShahidengr) [inline Bisal/assemblycodes/task1_2.s:25]: Once again, above comment is not matching the instruction

File: Bisal/codes/exercise1.c [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        //reverse string by swaping
+        char temp = str[i];
+        str[i] = str[n - i - 1];
+        str[n - i - 1] = temp;
+    }
+}
+
+int secondLargest(int arr[], int size) {
+    int first, second;
+    if (size < 2) {
```
Comment (UmerShahidengr) [inline Bisal/codes/exercise1.c:160]: Good check

File: Bisal/tests/.gitkeep [hunk 11]
```diff
@@ -0,0 +1 @@
+
```
Comment (UmerShahidengr) [inline Bisal/tests/.gitkeep:1]: An empty file found

File: Bisal/scripts_backup2024-07-04.tar.gz
Comment (UmerShahidengr) [inline Bisal/scripts_backup2024-07-04.tar.gz:1]: Remove this binary file, this should not be part of any PR

**Review by UmerShahidengr — CHANGES_REQUESTED** [review]

File: Bisal/assemblycodes/division_nonrestore.s [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    slli x4, x4, 1
+    
+    # Decrement loop counter
+    addi t0, t0, -1
+    bnez t0, loop      # Loop until t0 != 0
+    
+    # End of division, x3 contains quotient, x4 contains final remainder
+
+    # End of program
+    ecall
```
[the remark this answers is not in the document]
Comment (codewithbisal) [reply Bisal/assemblycodes/division_nonrestore.s:67]: I read the riscv assembly manual which told that its also a way to exit from spike and i wanted to try but when it fell into trap i talked to sir ubaid about it and he explained me that  changes might be required in linker script so i changed to host exit for now sorry i forgot to commit it again .

**Pull request closed without merging**